### PR TITLE
Fix #10197 - Spurious unit test failure in CommandLineInterfaceFixture

### LINKED ISSUE #10197: Spurious unit test failure
Issue overview
--------------
There is an occasional unit test failure like [this](https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-50b8a5511ce559e5175524b943c26cc5b99d712d-Win64-Windows-10-VisualStudio-16.html), where it is complaining it can't operate on `{build_folder}/tst/EnergyPlus/unit/in.epw` because it is in use.  I think the problem comes down to [some code around this line](https://github.com/NREL/EnergyPlus/blob/50b8a5511ce559e5175524b943c26cc5b99d712d/tst/EnergyPlus/unit/CommandLineInterface.unit.cc#L107), where it is trying to setup the in.idf and in.epw for CLI testing.  I guess if you run things in a bunch of threads, they can stomp on each other.  Could we give each individual test its own IDF/EPW so there is no conflict?

What's even weirder is that it's like the unit test....passes...but also doesn't?  I feel like it is getting the exception raised, which signals the test failed, but because the file already exists from a different test, it goes ahead and passes.  Here's the output from that link:

```
Note: Google Test filter = CommandLineInterfaceFixture.IdfOnly
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from CommandLineInterfaceFixture
unknown file: error: C++ exception with description "copy_file: The process cannot access the file because it is being used by another process.: "C:/ci/runs/clone_branch\weather/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw", "C:\ci\runs\clone_branch\build\tst\EnergyPlus\unit\in.epw"" thrown in SetUpTestSuite().
[ RUN ] CommandLineInterfaceFixture.IdfOnly
[ OK ] CommandLineInterfaceFixture.IdfOnly (362 ms)
[----------] 1 test from CommandLineInterfaceFixture (362 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (366 ms total)
[ PASSED ] 1 test.
[ FAILED ] 0 tests, listed below:

0 FAILED TESTS
[ FAILED ] CommandLineInterfaceFixture: SetUpTestSuite or TearDownTestSuite

1 FAILED TEST SUITE
```

What do you think @jmarrec ?

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

- Fix #10197 

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/CMakeLists.txt [hunk 1]
```diff
@@ -291,4 +291,7 @@ set(test_src
     CommandLineInterface.unit.cc
     main.cc
 )
+# For CommandLineInterface.unit.cc, make an in.idf / in.epw in the test directory so we can test the legacy CLI (applies to CTest which sets the current directory to this one)
+configure_file("${PROJECT_SOURCE_DIR}/weather/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw" "${CMAKE_CURRENT_BINARY_DIR}/in.epw" COPYONLY)
+configure_file("${CMAKE_CURRENT_SOURCE_DIR}/Resources/UnitaryHybridUnitTest_DOSA.idf" "${CMAKE_CURRENT_BINARY_DIR}/in.idf" COPYONLY)
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/CMakeLists.txt:296]: If you use ctest to run the tests, the in.epw / in.idf will already be there.

File: tst/EnergyPlus/unit/CommandLineInterface.unit.cc [hunk 2]
```diff
@@ -102,12 +102,24 @@ class CommandLineInterfaceFixture : public EnergyPlusFixture
...
 
-        auto inputWeatherFilePath = configured_source_directory() / "weather/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw";
-        fs::copy_file(inputWeatherFilePath, FileSystem::getAbsolutePath("in.epw"), fs::copy_options::skip_existing);
+        {
+            auto destPath = FileSystem::getAbsolutePath("in.epw");
+            if (!fs::is_regular_file(destPath)) {
+                auto inputWeatherFilePath = configured_source_directory() / "weather/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw";
+                fs::copy_file(inputWeatherFilePath, destPath, fs::copy_options::skip_existing);
+            }
+        }
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/CommandLineInterface.unit.cc:122]: Test existence before copying. This handles the case you run the `energyplusapi_tests` .exe directly, in which case the current directory is whichever directory you are currently in.

Comment (Myoldmopar) [reply tst/EnergyPlus/unit/CommandLineInterface.unit.cc:122]: OK, this seems sufficient.  If we are doing it via ctest, the file will already be there.  If running by the gtest binary, it likely won't be multithreaded, so this shouldn't race.  I suppose there may be a super weird case where this still causes a race, but it's going to be crazy rare.  This is fine.

File: tst/EnergyPlus/unit/CMakeLists.txt (hunk 1 above)
Comment (Myoldmopar) [reply tst/EnergyPlus/unit/CMakeLists.txt:296]: :+1:

**Review by Myoldmopar — APPROVED** [review]
This looks good.  I'll throw ctest -j 20 at it a few times and make sure :)

**Comment by Myoldmopar** [comment]
Hmm, I'm getting a runtime error when running via ctest, and not even using -j.  

```
$ ctest -VV -R WeatherFile_Missing
UpdateCTestConfiguration  from :/builds/r/DartConfiguration.tcl
   (---snip---)
test 2623
    Start 2623: EnergyPlusFixture.WeatherRunPeriod_WeatherFile_Missing

2623: Test command: /builds/r/Products/energyplus_tests "--gtest_filter=EnergyPlusFixture.WeatherRunPeriod_WeatherFile_Missing"
2623: Test timeout computed to be: 10000000
2623: Note: Google Test filter = EnergyPlusFixture.WeatherRunPeriod_WeatherFile_Missing
2623: [==========] Running 1 test from 1 test suite.
2623: [----------] Global test environment set-up.
2623: [----------] 1 test from EnergyPlusFixture
2623: [ RUN      ] EnergyPlusFixture.WeatherRunPeriod_WeatherFile_Missing
2623: /tst/EnergyPlus/unit/WeatherManager.unit.cc:1925: Failure
2623: Expected: WeatherManager::GetNextEnvironment(*state, Available, ErrorsFound) throws an exception of type std::runtime_error.
2623:   Actual: it throws nothing.
1/1 Test #2623: EnergyPlusFixture.WeatherRunPeriod_WeatherFile_Missing ...SIGTRAP***Exception:   0.25 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.30 sec

The following tests FAILED:
	2623 - EnergyPlusFixture.WeatherRunPeriod_WeatherFile_Missing (SIGTRAP)
Errors while running CTest
Output from these tests are in: /builds/r/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.

```

I did rerun ccmake and regenerate, just to make sure, but this persists.  I'll try to leave this build folder in place for testing if you want me to check anything, @jmarrec

**Comment by Myoldmopar** [comment]
Confirmed this now works fine, thanks @jmarrec, merging.

**Pull request merged**